### ISSUE TITLE
Add the ability to unsubscribe object listeners

### ISSUE THREAD

**Issue by schrockwell** [krisleech/wisper#132]
This feature request is similar to the global `Wisper.unsubscribe` mechanism added in #118.

There are use cases where a publisher has many subscribers that come and go. Subscribers should have the ability to unsubscribe themselves, because as it stands right now subscribers will just keep accumulating in memory after they are no longer needed.

I propose adding an `#unsubscribe` method to `Wisper::Publisher` as a complement to `#subscribe`. It would be used to remove object registrations from the `local_registrations` array. There would probably have to be a mutex added to maintain thread safety, since unsubscribing would require looping over `local_registrations`.

Taking it a step further, it would be cool if `#unsubscribe` could also remove block registrations. This scenario might "just work" due to the duck typing of registrations.

**Comment by krisleech** [comment]
Can you describe your use case a bit more please. Usually when you subscribe listeners the publishing objects are themselves are short lived, at least in web applications, where they only exist for the duration of a request. Therefore memory consumption is not an issue. 

I should note that this will, almost certainly, be a feature of Wisper 2.0, however when I have time to finish Wisper 2.0, I do not know.

**Comment by schrockwell** [comment]
In my case, the Wisper publisher is listening to a subscription on a Redis channel and rebroadcasting messages locally within the Ruby app. I opted against having many Redis clients, which would quickly eat up socket connections and resources. The Redis client is long-lived because the subscription blocks, so it lives on its own thread and it's listening throughout the duration of the process.

So, in a nutshell, it's a pub/sub system subscribed to a pub/sub system. :)

Let me know if you want me to draft up a PR to help move things along – it sounds like 2.0 needs some activity.

**Comment by jdickey** [comment]
Cute. That's cleaner than some of the Rube Goldberg tech I've jammed together for similar use cases :)

**Comment by jrochkind** [comment]
I have this need for a contemplated use of wisper too. 

My use case is a Rails engine that comes with some default subscriptions, but the implementing app may wish to unsubscribe them in order to replace with custom subscribers.

**Comment by tonywok** [comment]
@krisleech Saw 2.0 was released. :tada: Any plans on adding the ability to unsubscribe a listener from a publisher?

Looking at https://github.com/krisleech/wisper/blob/master/lib/wisper/publisher.rb, it seems like we could add an `#unsubscribe` that removes a listener from the publishers `local_registrations`. Or is it just a use case the library opts out of implementing due to some complexity I'm not seeing?

**Comment by krisleech** [comment]
> I should note that this will, almost certainly, be a feature of Wisper 2.0, however when I have time to finish Wisper 2.0, I do not know.

The "almost certainly" did not materialize, I was planning to refactor Wisper (well more like a rewrite) but I didn't get the opportunity. But... if you have some ideas about how it can be added to the existing codebase I would be happy to consider a PR. Your suggestion sounds possible, we could use `object_id` to find the listener to unsubscribe in `local_registrations`.

I would suggest `#unsubscribe` and `#unsubscribe!`, one which raises if the listener is not found.

**Comment by chr0n1x** [comment]
hey @krisleech - sorry to poke at this issue after it has been closed. I'm running into a scenario that's similar to @schrockwell's. The short version is that objects that have attached listeners cannot be `Marshal::dump`ed. Do you have any recommended ways around this? I stumbled across this issue after wondering if I could just detach the listeners before I marshal the object.

The fuller picture though: At a high-level, I have an object that has a listener subscribed e.g.

```ruby
wisper_poro.subscribe(MyCustomListener.new)
```

I'm trying to store this into cache via:

```ruby
Encrypt.encrypt(Marshal::dump(wisper_poro), opts)
```

But getting the following error:
```
irb(main):017:0> Encrypt.encrypt(Marshal::dump(wisper_poro))
TypeError: no _dump_data is defined for class Proc
	from (irb):17:in `dump'
```
